File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> freq = new HashMap<>();
        for(int num:nums){
            freq.put(num, freq.getOrDefault(num,0)+1);
        }
        List<List<Integer>> list = new ArrayList<>();
        for(int i=0; i<=nums.length; i++){
            list.add(new ArrayList<>());
        }
        for(Map.Entry<Integer, Integer> entry:freq.entrySet()){
            list.get(entry.getValue()).add(entry.getKey());
        }
        int[] res = new int[k];
        int ind=0;
        for(int i=list.size()-1; i>=0 && ind<k; i--){
            if(!list.get(i).isEmpty()){
                for(int ele:list.get(i)){
                    res[ind]=ele;
                    ind++;
                }
            }
        }
        return res;
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        map={}
        for num in nums:
            map[num] = map.get(num, 0)+1
        res = [[] for _ in range(len(nums)+1)]
        for key in map:
            res[map[key]].append(key)
        ans = []
        for i in range(len(res)-1, 0, -1):
            if k<=0:
                break
            ans.extend(res[i])
            k -= len(res[i])
        return ans